modules/actions/join.js

```diff
@@ -129,9 +129,27 @@ export function actionJoin(ids) {
             return 'not_adjacent';
         }
 
+        var i;
+
+        // All joined ways must belong to the same set of (non-restriction) relations.
+        // Restriction relations have different logic, below, which allows some cases
+        // this prohibits, and prohibits some cases this allows.
+        var sortedParentRelations = function (id) {
+            return graph.parentRelations(graph.entity(id))
+                .filter((rel) => !rel.isRestriction())
```

Comment:
I assume this is this code that already handles restriction relations?

https://github.com/openstreetmap/iD/blob/1ee45ee1f03f0fe4d452012c65ac6ff7649e229f/modules/actions/join.js#L162

I’m not sure if it’s a problem, but [connectivity relations](https://wiki.openstreetmap.org/wiki/Relation:connectivity) are intended to be modeled the same way as restriction relations. (Connectivity relations have a preset but not an editor, so currently you have to create one manually or morph a restriction relation into one using the raw tag editor.) Maybe `hasFromViaTo` would cover both relation types reasonably well?

Reply:
I think you're right, but to do that right would require more changes than I want to get into in this PR.

Reply:
To elaborate a bit: I want to avoid getting into modifying the code you linked to so that it prohibits joining connectivity relations in the same situations as restriction relations, due to i18n nuances. But it would be easy enough to just exempt them from the code I added. That would preserve the existing behavior for connectivity relations, while disabling problematic merges for the much more common relation types like route.

Reply:
Okay, okay, you convinced me to just implement full connectivity support. 😄 

<img width="831" alt="Screen Shot 2021-09-08 at 8 28 47 PM" src="https://user-images.githubusercontent.com/98601/132617724-bd785bab-3cd3-4839-b6b8-362eabeca35b.png">

Reply:
Next @1ec5 is going to ask for support for [type=manoeuvre](https://wiki.openstreetmap.org/wiki/Relation:manoeuvre) relations. 😆